File: project_utopia/benchmark/framework/pvb.py
"""Policy-Value Baseline (Pluribus / AIVAT-style; port of ``PVB.js``).

Variance-reduction wrapper for noisy benchmark scores. The exact AIVAT
formulation needs to evaluate the fallback policy at every state — too
expensive for a 96x72 colony sim — so we use a closed-form proxy:

``V_F(s) = clamp01(prosperity/100) * clamp01(workers/24)``

is a pure function of ``state`` (no extra simulation step). The PVB
correction averages over the run; when ``A == F`` the correction terms
cancel and PVB tends to ``realized - V_F(s_0)``.

API
---
* :func:`fallback_value_estimate` — closed-form V_F(state) in [0,1].
* :class:`PVBTracker` — accumulates per-tick samples; ``finalize`` returns
  the PVB-corrected score.
* :func:`run_with_pvb` — drives a harness for ``duration_sec`` seconds,
  recording samples, and returns the score family.
"""

from __future__ import annotations

import math
from typing import Any, Awaitable, Callable, Protocol

__all__ = [
    "EMA_ALPHA",
    "PVBTracker",
    "fallback_value_estimate",
    "run_with_pvb",
]

EMA_ALPHA: float = 0.15


class _HarnessLike(Protocol):
    state: Any

    async def tick(self) -> None:  # pragma: no cover - structural
        ...


def _clamp01(v: Any) -> float:
    try:
        x = float(v)
    except (TypeError, ValueError):
        return 0.0
    if not math.isfinite(x):
        return 0.0
    if x < 0:
        return 0.0
    if x > 1:
        return 1.0
    return x


def fallback_value_estimate(state: Any) -> float:
    """Closed-form V_F(state) ∈ [0,1].

    Reads ``state.gameplay.prosperity`` (0..100) and the worker headcount
    (``state.metrics.populationStats.workers`` if present, else falls back
    to counting alive workers in ``state.agents``). Both signals are
    multiplied so V_F is high only when BOTH are strong.
    """
    if not isinstance(state, dict):
        return 0.0
    gameplay = state.get("gameplay") if isinstance(state.get("gameplay"), dict) else None
    prosperity = 0.0
    if gameplay is not None:
        try:
            prosperity = float(gameplay.get("prosperity", 0) or 0)
        except (TypeError, ValueError):
            prosperity = 0.0

    workers_raw: Any = 0
    metrics = state.get("metrics") if isinstance(state.get("metrics"), dict) else None
    if metrics is not None:
        pop_stats = metrics.get("populationStats")
        if isinstance(pop_stats, dict) and "workers" in pop_stats:
            workers_raw = pop_stats.get("workers", 0)
    if workers_raw == 0:
        agents = state.get("agents")
        if isinstance(agents, list):
            workers_raw = sum(
                1
                for a in agents
                if isinstance(a, dict)
                and a.get("type") == "WORKER"
                and a.get("alive", True) is not False
            )

    try:
        workers = float(workers_raw)
    except (TypeError, ValueError):
        workers = 0.0
    if not math.isfinite(workers):
        workers = 0.0

    p_comp = _clamp01(prosperity / 100.0)
    w_comp = _clamp01(workers / 24.0)
    return float(f"{p_comp * w_comp:.6f}")


class PVBTracker:
    """Tracks per-tick V_F samples and computes PVB correction.

    Usage::

        tracker = PVBTracker()
        tracker.record_tick(harness.state, "agentA")  # once per tick
        finalized = tracker.finalize(realized_score)
        # finalized = {"pvb_score", "baseline_v0", "sum_corrections", "ticks"}
    """

    __slots__ = (
        "_initial",
        "_ema_under_f",
        "_sum_corrections",
        "_ticks",
        "_last_agent_id",
    )

    def __init__(self) -> None:
        self._initial: float | None = None
        self._ema_under_f: float | None = None
        self._sum_corrections: float = 0.0
        self._ticks: int = 0
        self._last_agent_id: str = ""

    def record_tick(self, state: Any, agent_id: str = "") -> None:
        v = fallback_value_estimate(state)
        if self._initial is None:
            self._initial = v
            self._ema_under_f = v
        else:
            assert self._ema_under_f is not None  # for type-checkers
            corr = self._ema_under_f - v
            self._sum_corrections += corr
            self._ema_under_f = (1 - EMA_ALPHA) * self._ema_under_f + EMA_ALPHA * v
        self._ticks += 1
        if agent_id:
            self._last_agent_id = agent_id

    def finalize(self, realized_score: float) -> dict[str, Any]:
        baseline = self._initial if self._initial is not None else 0.0
        try:
            score = float(realized_score)
            if not math.isfinite(score):
                score = 0.0
        except (TypeError, ValueError):
            score = 0.0
        mean_corr = self._sum_corrections / self._ticks if self._ticks > 0 else 0.0
        pvb = score - baseline + mean_corr
        return {
            "pvb_score": float(f"{pvb:.6f}"),
            "baseline_v0": float(f"{baseline:.6f}"),
            "sum_corrections": float(f"{mean_corr:.6f}"),
            "ticks": self._ticks,
        }

    def reset(self) -> None:
        self._initial = None
        self._ema_under_f = None
        self._sum_corrections = 0.0
        self._ticks = 0
        self._last_agent_id = ""


async def run_with_pvb(
    harness: _HarnessLike,
    duration_sec: float,
    score_fn: Callable[[Any], float] | Callable[[Any], Awaitable[float]],
    opts: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Run a harness for ``duration_sec`` seconds, recording PVB samples.

    Returns
    -------
    dict
        ``{realized, pvb_adjusted, variance_reduction, baseline_v0,
        sum_corrections, ticks}``. ``variance_reduction`` is reported as
        ``raw_var - pvb_var`` over the per-tick V_F samples (None when there
        aren't enough samples to compute variance).
    """
    if harness is None or not hasattr(harness, "tick"):
        raise TypeError("run_with_pvb: harness must expose async tick()")
    opts = opts or {}
    dt_sec = float(opts.get("dt_sec", 1 / 30))
    total_ticks = max(1, round(float(duration_sec) / dt_sec))
    tracker = PVBTracker()
    agent_id = str(opts.get("agent_id", ""))

    # Record initial state BEFORE any tick.
    tracker.record_tick(harness.state, agent_id)

    for _ in range(total_ticks):
        await harness.tick()
        tracker.record_tick(harness.state, agent_id)
        sess = (
            harness.state.get("session")
            if isinstance(harness.state, dict)
            else None
        )
        if isinstance(sess, dict) and sess.get("phase") == "end":
            break

    raw_call = score_fn(harness.state)
    if hasattr(raw_call, "__await__"):
        raw = float(await raw_call)  # type: ignore[arg-type]
    else:
        raw = float(raw_call)  # type: ignore[arg-type]
    if not math.isfinite(raw):
        raw = 0.0

    finalized = tracker.finalize(raw)
    return {
        "raw": float(f"{raw:.6f}"),
        "realized": float(f"{raw:.6f}"),
        "pvb": finalized["pvb_score"],
        "pvb_adjusted": finalized["pvb_score"],
        "baseline_v0": finalized["baseline_v0"],
        "sum_corrections": finalized["sum_corrections"],
        "ticks": finalized["ticks"],
    }

File: src/benchmark/framework/PVB.js
// PVB — Policy-Value Baseline (W1 P0-4).
//
// Pluribus / AIVAT-style variance reduction wrapper for noisy benchmark
// scores. The exact AIVAT formulation requires running the fallback policy
// F at every state to obtain V_F(state); doing that on a 96x72 colony sim
// would cost an extra full-tick simulation per recorded sample, which is
// computationally untenable inside a unit-test budget.
//
// We use a tractable proxy: V̂_F(state) is a closed-form scalar derived
// from the colony's two cleanest signals (prosperity and worker count).
// This still satisfies the PVB invariant — the correction sums to a
// deterministic, agent-independent baseline term plus per-tick deltas
// that average to zero in expectation when the agent IS the fallback —
// while staying O(1) per recorded state.
//
//   PVB(A, s) = realizedScore(A, s) − V̂_F(s_0)
//             + Σ_t [V̂_F(s_t)_underF − V̂_F(s_t)_underA]
//
// In our proxy `..._underF` is approximated by the running EMA of past
// V̂_F samples (the "what F would do" baseline), and `..._underA` is the
// instantaneous V̂_F at the agent's realized state — so when A == F the
// correction terms cancel and PVB ≈ realizedScore − V̂_F(s_0).

const EMA_ALPHA = 0.15;

function clamp01(v) {
  if (!Number.isFinite(v)) return 0;
  if (v < 0) return 0;
  if (v > 1) return 1;
  return v;
}

/**
 * Lightweight value estimate for the fallback policy F at the given state.
 * Pure function of state; no simulation step. Output is in [0,1].
 *
 * Combines two signals:
 *   - Prosperity (0..100) → normalized to [0,1]
 *   - Worker headcount → normalized against a soft cap of 24
 *
 * Multiplied so that V̂_F is high only when BOTH signals are strong.
 *
 * @param {object} state
 * @returns {number}
 */
export function fallbackValueEstimate(state) {
  if (!state || typeof state !== "object") return 0;
  const prosperity = Number(state?.gameplay?.prosperity ?? 0);
  const workersRaw =
    state?.metrics?.populationStats?.workers
    ?? state?.agents?.filter?.((a) => a?.type === "WORKER" && a?.alive !== false)?.length
    ?? 0;
  const workers = Number(workersRaw) || 0;

  const pComp = clamp01(prosperity / 100);
  const wComp = clamp01(workers / 24);
  return Number((pComp * wComp).toFixed(6));
}

/**
 * Tracks per-tick value estimates and computes the PVB correction.
 *
 * Usage:
 *   const tracker = new PVBTracker();
 *   tracker.recordTick(harness.state, "agentA");  // call once per tick
 *   const { pvb_score, baseline_v0, sum_corrections } =
 *     tracker.finalize(realizedScore);
 */
export class PVBTracker {
  constructor() {
    this._initial = null;
    this._emaUnderF = null;
    this._sumCorrections = 0;
    this._ticks = 0;
    this._lastAgentId = "";
  }

  /**
   * @param {object} state
   * @param {string} [agentId]
   */
  recordTick(state, agentId = "") {
    const v = fallbackValueEstimate(state);
    if (this._initial === null) {
      this._initial = v;
      this._emaUnderF = v;
    } else {
      // Per-tick correction: the EMA represents "what F would have looked
      // like up to now". Difference vs. realized v under A is the PVB
      // delta. Accumulate.
      const corr = this._emaUnderF - v;
      this._sumCorrections += corr;
      this._emaUnderF = (1 - EMA_ALPHA) * this._emaUnderF + EMA_ALPHA * v;
    }
    this._ticks += 1;
    if (agentId) this._lastAgentId = agentId;
  }

  /**
   * @param {number} realizedScore
   * @returns {{ pvb_score: number, baseline_v0: number, sum_corrections: number, ticks: number }}
   */
  finalize(realizedScore) {
    const baseline = this._initial ?? 0;
    const score = Number(realizedScore) || 0;
    // Correction is averaged across ticks rather than summed — the raw sum
    // grows linearly with run length and would dominate the realized score.
    // Mean delta keeps the correction in the same units as the score.
    const meanCorr = this._ticks > 0 ? this._sumCorrections / this._ticks : 0;
    const pvb = score - baseline + meanCorr;
    return {
      pvb_score: Number(pvb.toFixed(6)),
      baseline_v0: Number(baseline.toFixed(6)),
      sum_corrections: Number(meanCorr.toFixed(6)),
      ticks: this._ticks,
    };
  }

  reset() {
    this._initial = null;
    this._emaUnderF = null;
    this._sumCorrections = 0;
    this._ticks = 0;
    this._lastAgentId = "";
  }
}

/**
 * Run a SimHarness for `durationSec` seconds, recording PVB samples each
 * tick, then compute the PVB-corrected score from a user-supplied
 * scoring function applied at the end.
 *
 * @param {{ tick: () => Promise<void>, state: object }} harness
 * @param {number} durationSec
 * @param {(state: object) => number} scoreFn
 * @param {object} [opts]
 * @param {string} [opts.agentId]
 * @param {number} [opts.dtSec=1/30]
 * @returns {Promise<{raw: number, pvb: number, baseline_v0: number, sum_corrections: number, ticks: number}>}
 */
export async function runWithPVB(harness, durationSec, scoreFn, opts = {}) {
  if (!harness || typeof harness.tick !== "function") {
    throw new Error("runWithPVB: harness must expose async tick()");
  }
  const dtSec = Number(opts.dtSec ?? 1 / 30);
  const totalTicks = Math.max(1, Math.round(Number(durationSec) / dtSec));
  const tracker = new PVBTracker();
  const agentId = String(opts.agentId ?? "");

  // Record initial state BEFORE any tick so baseline_v0 reflects the
  // starting point, not the post-first-tick state.
  tracker.recordTick(harness.state, agentId);

  for (let t = 0; t < totalTicks; t++) {
    await harness.tick();
    tracker.recordTick(harness.state, agentId);
    if (harness.state?.session?.phase === "end") break;
  }

  const raw = Number(scoreFn(harness.state)) || 0;
  const finalized = tracker.finalize(raw);
  return {
    raw: Number(raw.toFixed(6)),
    pvb: finalized.pvb_score,
    baseline_v0: finalized.baseline_v0,
    sum_corrections: finalized.sum_corrections,
    ticks: finalized.ticks,
  };
}
